### DISCUSSION TITLE
DAOIP-5: Grants Management

### DISCUSSION THREAD

**Discussion by kevinrolsen** [metagov/daostar#51] [General]
Contributors: Tyler Sully, Kevin Olsen, Riccardo Lamanna, Aditya Anand, Balazs Nemethi, Apoorv Nandan, Tina He, and Joshua Tan.

This is a proposed grants management standard authored by the DAOstar Grants Management working group, including members of Gitcoin, DaoLens, OpenQ, and Station. It is a work-in-progress, but before we submit this as a DAOIP we would appreciate any comments or suggestions from the community.

_To facilitate comments and discussion, we have italicized fields and decisions that we feel may require more discussion and development._

# Simple summary
Common interfaces and structures for grants management.

# Motivation
Grant programs have become a popular tool for DAOs and blockchain projects to accomplish tasks and build infrastructure. The explosion in the popularity of grants programs has been accompanied by an explosion of different approaches for managing grants. The diversity of approaches has led to innovative approaches for grants management but has also fragmented capital and reduced transparency. 

This specification lays out a common set of on- and off-chain interfaces for grant programs looking to deploy funding and for projects looking to receive funding. It is intended to increase interoperability between grants tooling, standardize interactions with voting and payout systems, improve the transparency of grant programs, and ultimately make grant programs more efficient for funders, administrators, and grantees.

# Specification
The key words “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”, “SHOULD NOT”, “RECOMMENDED”, “MAY”, and “OPTIONAL” in this document are to be interpreted as described in RFC 2119.

This specification is organized into five main components:
1. Grant system administration
2. Grant pool
3. Application
4. Payouts
5. Voting

While eligibility and compliance flows (including KYC) are important in grants systems, we decided to not include them within the scope of the current specification.

## Grants system
The grants system module is intended to be a thin wrapper over one or more grant pools (see next section) that manages basic metadata and declares a relationship (e.g. of ownership) between a given DAO and a given grant system. Roles and permissions to manage a given grant system are not captured at this stage.

Grants System JSON-LD Schema
```json
{
    "@context": "http://www.daostar.org/schemas",
    "type": "grant-system",
    "id": "<a generic id of the DAO operating the grants system, e.g. its CAIP-10 contract address>",
    "daoURI": "<daoURI, see DAOIP-2 / EIP-4824>",
    "pool_id": "<id of the grant object>",
}
```

## Grant pool
A grant pool is funding locked in a smart contract or other fundable object with the intention of being used to pay out grants. It is essentially a subDAO with off-chain metadata describing the purpose/process of the grant pool and on-chain methods for controlling the flow of funds from the pool.

The proposed grant pool design is designed to enable clear reporting of grant pool and application status while remaining flexible with respect to many different program designs including Gitcoin-style “rounds”, rolling application/funding period, custom refund logic, etc.

Grant Pool JSON-LD Schema
```json
{
    "@context": "http://www.daostar.org/schemas",
    "type": "grant-pool",
    "name": "<name of the Grant Pool>",
    "description": "<description>",
    "grantSystemURI": "<URI>",
    "tokenAllowList": ["<address>"], 
    "funderAllowlist": ["<address>"],
    "fundingActive": "<boolean>",
    "appliationActive": "<boolean>",
    "applicationStartDate": "<date>",
    "applicationEndDate": "<date>",
    "applicationsURI": "<URI>",
    "lockedFunds": [
	{"tokenAddress": "<address>",
	"amount": "<integer>"}],
}
```

_We have not yet decided on the final on-chain grant pool. However, we have listed some of the potentially necessary on-chain methods below:_

* init(name)
	* [TBD] will likely include setters for governance and payout method
* addTokenAllowList
* removeTokensAllowList
* deposit({tokenAddress: amount})
	* Restricted to members of the allowList
	* Revert if fundingActive = false
* setFundingActive (boolean)
	* Only callable by grant contract’s governance
* setApplicationActive (boolean)
	* Only callable by grant contract’s governance
* setAllowList ({address: boolean})
	* Only callable by grant contract’s governance
* refund({address, tokens})
	* Only callable by grant contract’s governance
	* Disburses all funds but doesn’t specify exactly how funds are distributed
	* Should revert if not all funds are distributed
	* Sets fundingActive and applicationActive to false
* apply()
	* Anyone can call this function
	* Revert if applicationActive = false
* setEligible({address: boolean})
* payout(applicationId, {tokenAddress: amount},  payoutAddress)
	* Only callable by grant contract’s governance
	* Pays out funds from pool

## Applications
Projects (see below) are independent of any particular grant program, but when a project applies for a grant from a particular grant pool, it is called an application to that pool. The following schema defines `applicationsURI`, which when queried returns all pending and past applications to a grant pool.

Applications URI JSON-LD Schema

```json
{
    "@context": "http://www.daostar.org/schemas",
    "type": "grant-application",
    "name": "<name of the Grant Pool>",
    "applications": [
      {
       "projectURI": "<URI>",
       "eligible": "<boolean>",
       "payout": [
	{"tokenAddress": "<address>",
       "payoutAddress": "<address>",
	 "amount": "<integer>"}]
      } 
  ]
}
```

## Projects
Projects are proposals that may be submitted to a grant pool or program.

Grant applications to a grant pool are very similar to generic proposals to a DAO. However, individual projects often apply to multiple grant pools or multiple times to a single grant pool, for example multiple Gitcoin rounds or a co-funding round for a public good supported by many DAOs). For this purpose, we define projects separately from applications.

_An open question in this section is how we should attach the history of a project or application (see “previousSuccessfulProposalLinks” below). Secondary question here is how to attach history, or if we want to attach history to the project and if we want to expose that history when it applies to a grant pool._

Project URI JSON-LD Schema
```json
{
    "@context": "http://www.daostar.org/schemas",
    "type": "grant-project",
    "id": "<id>",
    "name": "<name>",
    "description": "<description>",
    "website": "<URL>",
    _"credentials": "<TBD>",_
    "previousSuccessfulProposalLinks": [],
    "createdAt": "<Unix timestamp>",
    "status": "<text>",
    "fundsAsked": "<id>",
    "milestones": ["<deadlines and the like>"],
    "links": [],
    "roles": ["<access control, milestone reviewers, etc.>"],
    "walletAddress": [],
    "email": [],
    _"logoImg": [],_
    _"logoImgData": [],_
    "members": [ {"type": "EthereumAddress", "id": "JXxsnO9yqahgfwXg_C1nZ", "text": "Aman Sehrawat", "email": "a@daolens.scom"}],
    _"smartContractCaller": "<what does this mean exactly?>"_
}
```

## Payouts
Once approved, there are many different ways that a grant pool may emit a payout to the project, including direct transfers, milestone-based payouts, time-based payouts, escrow, and other mechanisms. To support such payouts, an on-chain grant pool satisfying DAOIP-# MAY implement the following payout interface:

```solidity
/// @title Payout required data/ funtions
/// @dev Would be required to be implemented by any payout strategy 

interface PayoutInterface {

    /// @notice Pool Address
    address round;

    /// @notice checks if caller is round owner. 
    modifier isRoundOwner();

    /// @notice Used to update distribution which are needed for payout to happen
    /// as well IPFS hash to the actual dataset to make it easy for other to verify on
    /// how the distribution was generated.
    /// The actual structure of of the encoding would be defined by the payout
    /// implementation as every implementations would have distinct data it would want to store
    function updateDistribution(bytes[] calldata _encodedDistribution) external virtual isRoundOwner;

}

    /// @notice Used to trigger the actutal distribution to the grantee
    /// The actual distribution would be implemented by payout implementation
    function distribute(bytes[] calldata _encodedPayoutData) external virtual;

    /// @notice Used to withdraw funds in scenarios where airdrops haven’t been claimed.
  Function withdraw() external virtual isRoundOwner;
```

## Voting
A on-chain grant pool will often be accompanied by a particular method for deciding projects, e.g. some form of voting. Where voting is encoded in a smart contract, a grant pool satisfying DAOIP-# MAY implement the following interface:

```solidity
/// @title VotingInterface required functions / fields
/// @dev Would be required to be implemented by any voting strategy 

interface VotingInterface {

    /// @notice Round Address
    address round;

    /// @notice Expects an array of encoded votes allowing single/ multiple votes 
    /// to be casted. The actual structure of the vote will be defined by the voting
    /// strategy implementation. 
    /// @dev Ensure votes cannot be casted once the pool has closed or become inactive. 
    /// This can be done by either
    /// - ensuring this function can be invoked only by pool 
    /// - using the round address to define when a vote can be casted
    function vote(bytes[] calldata _encodedVotes, address _voter) external virtual payable;

}
```

# Rationale
The specification above was based on an extended study of different grant programs and grant tooling for DAOs and other Web3 projects. We have reproduced some of the diagrams associated with that research below. We will extend this section in future versions of this draft.

![gitcoin_diagram](https://user-images.githubusercontent.com/9309884/224125274-1df18965-554f-4540-ba4f-947c67887f15.jpg)

Gitcoin use-case for DAOIP-#.

![openq_diagram](https://user-images.githubusercontent.com/9309884/224124232-eaebc6c2-c250-478a-aebf-118919fa9a78.jpg)

OpenQ use-case for DAOIP-#.

![daolens_diagram](https://user-images.githubusercontent.com/9309884/224125176-b4c5d793-2f09-4bec-b062-243208df570c.jpg)

DaoLens use-case for DAOIP-#.

![solana_diagram](https://user-images.githubusercontent.com/9309884/224124275-eed97970-f076-4c04-ac7a-9ae84536d467.jpg)

Solana use-case for DAOIP-#.

![grants_legend](https://user-images.githubusercontent.com/9309884/224125337-8da09025-bb1e-44d8-a75d-dc4713ab1388.jpg)

Legend.

# Copyright
Copyright and related rights waived via [CC0](https://creativecommons.org/share-your-work/public-domain/cc0/).

**Comment by amanwithwings** [comment]
Hey all! Posting a couple of questions from my initial reading of the standard: 

1. Why assume grants pool will be DAO governed? The big grant programs - polygon, Avalanche, NEAR, Arbitrum etc. aren't DAO governed (as far as I know). Is it our expectation that this centralization will (need to) change? 

2. Should applicantAllowList, token allowlist be fundamental properties of the grant pool? For eg, locked funds from a grant round may carry over to the next round (assuming the grant pool is still the same contract between rounds), but allowlist may change each round- wouldn't this make things harder, for eg to track payouts from each round?

Also tagging @thelastjosh to keep track.

**Reply by thelastjosh** [reply]
> Why assume grants pool will be DAO governed? The big grant programs - polygon, Avalanche, NEAR, Arbitrum etc. aren't DAO governed (as far as I know). Is it our expectation that this centralization will (need to) change?

This standard doesn't actually assume that grant pools are DAO governed. We do expect any grants system administrator to report daoURI, which typically requires that they maintain control over some contract address that is registered with daoURI. Note that a foundation can have a DAO without being governed by that DAO. In this case, the DAO is just a smart contract that the foundation controls that it uses to interface with on-chain tooling such as a grants protocol.

Part of the idea here is to force big foundations (and other entities) to have an on-chain entity, i.e. a DAO, to interact with this tooling. It does not change the centralization or governance of the foundation or grant program.

> Should applicantAllowList, token allowlist be fundamental properties of the grant pool? For eg, locked funds from a grant round may carry over to the next round (assuming the grant pool is still the same contract between rounds), but allowlist may change each round- wouldn't this make things harder, for eg to track payouts from each round?

You're right that it may make tracking payouts harder. The reason we want `funderAllowList` as a basic field in the standard is for KYC / eligibility / compliance reasons. I.e. Gitcoin doesn't want to allow certain problematic addresses (e.g. addresses subject to AML enforcement) to donate, because effectively poison the grant pool.

We might end up removing `applicantAllowList`; this field was a bit of a last-minute addition because the use-case for it overlaps with eligibility.

`tokenAllowList` was recommended by Gitcoin based on their experience for severals reasons, e.g. because otherwise you may end up with a lot of spam tokens. And it's much harder to organize these grant pools if there's a bajillion different tokens in them.

**Reply by nkvdmax** [reply]
good

**Comment by free2ride19** [comment]
good to be here ... further information required for adjustability

**Comment by thelastjosh** [comment]
The new grants standard is (finally) up and officially published! 🚀 
Take a look: https://github.com/metagov/daostar/blob/main/DAOIPs/daoip-5.md
Will still be rolling some edits into it, but it's ready for an initial round of community feedback. Looking forward to it.

**Comment by thelastjosh** [comment]
In conversation with Mahesh: it would be nice to have social links (Twitter, GitHub, etc.) as part of the project profile.

Also, it might be nice to have "team members" attached to a project (or to a project), who can share reputation in some way.

**Comment by ankeliu** [comment]
The schemas actually align quite well to what we use at the Stellar Community Fund (just different naming), I think for the Cartographer Syndicate registry we may need to add "Funding Organization(s)" and "Ecosystem(s)" either per grant system or grant pool (grant pool may be more inclusive). Other things on projects that would be good to have added is their website and github. The payout schema is currently optional but that's the most important thing imo, the payment itself with a status indicator that it's paid or approved. In addition to what's currently listed in the payout schema, I'd also add a currency denominator in the payout schema (for instance, awards may be denominated in USD but paid out in crypto) and it also should have date field so it's possible to calculate roughly the USD value of the payment.

**Comment by slice-amandata** [comment]
@thelastjosh  hey i was also a contributor from daolens team.can you add my name in contributors.

Github Repo Link from daolens - https://github.com/daolens/funding-protocol-contracts

**Comment by free2ride19** [comment]
Payout/No payout:
No payment made yet to syndicate.
"Great to hear that the schemas align with the Stellar Community Fund's
existing structure! For the Cartographer Syndicate registry, considering
the addition of:

- Funding Organization(s)
- Ecosystem(s) (per grant pool or grant system)

Makes sense.

Also, suggesting additional project fields:

- Website
- GitHub repository

And enhancements to the payout schema:

- Currency denominator (e.g., USD, crypto)
- Date field for calculating USD value
- Status indicator (paid/approved)

These modifications will provide valuable clarity and organization to the
grant and payout processes.

Looking forward to implementing these adjustments and improving the
Cartographer Syndicate registry!"

On Monday, November 4, 2024, slice-amandata ***@***.***>
wrote:

> @thelastjosh <https://github.com/thelastjosh> hey i was also a
> contributor from daolens team.can you add my name in contributors.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/metagov/daostar/discussions/51#discussioncomment-11142807>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AXV67YZQ475L22RCGTLI75TZ65263AVCNFSM6AAAAABK2GOB6CVHI2DSMVQWIX3LMV43URDJONRXK43TNFXW4Q3PNVWWK3TUHMYTCMJUGI4DANY>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>


-- 
null
